crates/ironrdp-displaycontrol/src/client.rs

```diff
@@ -22,29 +59,12 @@ impl DvcProcessor for DisplayControlClient {
     }
 
     fn process(&mut self, _channel_id: u32, payload: &[u8]) -> PduResult<Vec<DvcMessage>> {
-        // TODO: We can parse the payload here for completeness sake,
-        // in practice we don't need to do anything with the payload.
-        debug!("Got Display PDU of length: {}", payload.len());
-        Ok(Vec::new())
+        let caps = DisplayControlCapabilities::decode(&mut ReadCursor::new(payload))?;
+        debug!("received {:?}", caps);
+        (self.on_capabilities_received)(caps)
     }
 }
 
 impl DvcClientProcessor for DisplayControlClient {}
 
-impl DisplayControlClient {
-    pub fn new() -> Self {
-        Self {}
-    }
-
-    /// Fully encodes a [`MonitorLayoutPdu`] with the given monitors.
-    pub fn encode_monitors(&self, channel_id: u32, monitors: Vec<MonitorLayoutEntry>) -> PduResult<Vec<SvcMessage>> {
-        let pdu: DisplayControlPdu = DisplayControlMonitorLayout::new(&monitors)?.into();
-        encode_dvc_messages(channel_id, vec![Box::new(pdu)], ChannelFlags::empty())
-    }
-}
-
-impl Default for DisplayControlClient {
-    fn default() -> Self {
-        Self::new()
-    }
-}
+type OnCapabilitiesReceived = Box<dyn Fn(DisplayControlCapabilities) -> PduResult<Vec<DvcMessage>> + Send + Sync>;
```

Comment:
I think we could drop `Sync` trait here

Reply:
We can't currently, because we have

```
impl DvcClientProcessor for DisplayControlClient {}
pub trait DvcClientProcessor: DvcProcessor {}
pub trait DvcProcessor: AsAny + Send + Sync {
    /// The name of the channel, e.g. "Microsoft::Windows::RDS::DisplayControl"
    fn channel_name(&self) -> &str;

    /// Returns any messages that should be sent immediately
    /// upon the channel being created.
    fn start(&mut self, channel_id: u32) -> PduResult<Vec<DvcMessage>>;

    fn process(&mut self, channel_id: u32, payload: &[u8]) -> PduResult<Vec<DvcMessage>>;

    fn close(&mut self, _channel_id: u32) {}
}
```

I traced the lineage of the requirement of `Sync` for `DvcProcessor` back to the requirement on `DynamicVirtualChannel` in a @CBenoit PR from several months back: https://github.com/Devolutions/IronRDP/pull/183/files#diff-34003f425460df2c4178a4c2a9b61ca57ece7544ed13d42975dfa8f8e016f224R37

Unless we can recall a specific reason for it, maybe we can remove it here?

Reply:
Sounds good to me! @ibeckermayer